### PR TITLE
fixes #20954; bounchecks for len(toOpenArray()) [backport]

### LINKED ISSUE #20954: Missing bounds check for `len(toOpenArray..)`
### Description

```nim
var v: array[10, int]

echo len(toOpenArray(v, 20, 30))
```

`toOpenArray` should in this case generate an out-of-bounds defect since 20 > 10

### Nim Version

devel, 1.6.8

### Current Output

_No response_

### Expected Output

_No response_

### Possible Solution

_No response_

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by ringabout**
fixes #20954

### CODE REVIEW

File: compiler/ccgexprs.nim [hunk 1]
```diff
@@ -1879,9 +1879,13 @@ proc genArrayLen(p: BProc, e: PNode, d: var TLoc, op: TMagic) =
     # Bug #9279, len(toOpenArray()) has to work:
     if a.kind in nkCallKinds and a[0].kind == nkSym and a[0].sym.magic == mSlice:
       # magic: pass slice to openArray:
+      var m: TLoc
```
Comment (arnetheduck) [inline compiler/ccgexprs.nim:1882]: there's a `genSlice` that potentially could be used, but from the looks of it, it also lacks bounds checks

Comment (ringabout) [reply compiler/ccgexprs.nim:1882]: I don't know. In this case, there is no need to generate slice. A bound check should suffice.

Comment (arnetheduck) [reply compiler/ccgexprs.nim:1882]: so I've noticed, ie the cgen has special slice handling all over, often duplicated, and each place is slightly different leading to bugs like this - collecting it all in one place seems like a worthwhile effort - basically, treating everything as a reified openarray would significantly simplify the flow and thus make it less error-prone.

**Review by arnetheduck — APPROVED** [review]
thanks!

**Pull request merged**